src/Main.elm

```diff
@@ -25,40 +143,939 @@ init =
+            , case daysString of
+                Just day ->
+                    text <| pluralize day "day"
+
+                Nothing ->
+                    text ""
+            ]
+        ]
+
+
+formWrapper : List (Html Msg) -> Html Msg
+formWrapper children =
+    Html.form [ Attr.class "flex flex-col" ]
+        children
+
+
+pluralize : String -> String -> String
+pluralize int str =
+    if (int |> String.toInt |> Maybe.withDefault 0) == 1 then
+        str
+
+    else
+        str ++ "s"
+
+
+viewInfoIcon : String -> Bool -> Html Msg
+viewInfoIcon txt showTooltip =
+    Html.div [ Attr.class "ml-1 w-[17px] h-[17px] text-center text-sm text-white rounded-full bg-blue-400 cursor-pointer relative", onMouseEnter ShowTooltip, onMouseLeave ShowTooltip ]
+        [ text "?"
+        , if showTooltip then
+            Html.p [ Attr.class "absolute -top-1 left-[20px] w-[420px] rounded text-xs bg-black p-1 flex " ] [ text txt ]
+
+          else
+            text ""
+        ]
+
+
+viewInputText : { id : Ids, value : String, overrideClass : Maybe String } -> Html Msg
+viewInputText { id, value, overrideClass } =
+    Html.input
+        [ Attr.class "ml-4 rounded-sm border-0 py-1 px-2 h-10 text-gray-900 shadow-sm h-[30px] text-sm ring-transparent focus:ring-transparent"
+        , Attr.class
+            (case overrideClass of
+                Just override ->
+                    override
+
+                Nothing ->
+                    ""
+            )
+        , Attr.type_ "text"
+        , Attr.id (fromIdToString id)
+        , onInput (id |> OnInputChange)
+        , Attr.value value
+        ]
+        []
+
+
+viewContactDetails : FormData -> { isInherited : Bool, prefMethod : PreferredContactMethod } -> Bool -> Html Msg
+viewContactDetails { inputFields } { isInherited, prefMethod } shouldShowDropdown =
+    Html.section [ Attr.class "relative" ]
+        [ if isInherited then
+            viewDisablePageOverlay
+
+          else
+            text ""
+        , Html.p [ Attr.class "mb-10" ] [ text "User's contact details" ]
+        , formWrapper
+            [ Html.div [ Attr.class "flex mb-10 items-center" ]
+                [ Html.div [] [ text "Choose preferred contact method" ]
+                , viewPreferredMethodDropdown { prefMethod = prefMethod, shouldShowDropdown = shouldShowDropdown }
+                ]
+            , viewFieldset
+                { id = EmailId
+                , daysString = Nothing
+                , requiredField = prefMethod == Email
+                , textValue = "Email"
+                , inputValue = inputFields |> Dict.get "email" |> Maybe.withDefault ""
+                , maxWidthFieldClass = "w-[132px]"
+                , overrideClass =
+                    if prefMethod == Email then
+                        Just "rounded-t-none ring-blue-400 ring-1"
+
+                    else
+                        Nothing
```

Comment:
Q: what is the reason not to refactor this away? Seems like it is duplicated aver and over

Reply:
I don't think 3 conditionals are duplicate, I need those overrides only in specific cases, and because **viewFieldset** is handling all inputs, needed the way to distinguish them.